### ISSUE TITLE
`ScalarNoise::computeMaxPower` is incorrect when there is no background noise module

### ISSUE THREAD

**Issue by edlongman** [inet-framework/inet#700]
The way that compute max power and compute min power are calculated is not correct. 

For example: 
```
    simtime_t startTime = SimTime(10, SimTimeUnit::SIMTIME_MS);
    simtime_t endTime = SimTime(11, SimTimeUnit::SIMTIME_MS);
    inet::Hz centerFrequency = inet::MHz(1000);
    inet::Hz bandwidth = inet::kHz(1);
    std::map<simtime_t, inet::W> powerChangesBase;
    powerChangesBase.insert(std::pair<simtime_t, inet::W>(
            SimTime(5, SimTimeUnit::SIMTIME_MS), inet::W(5) ));
    powerChangesBase.insert(std::pair<simtime_t, inet::W>(
            SimTime(8, SimTimeUnit::SIMTIME_MS), inet::W(3) ));
    powerChangesBase.insert(std::pair<simtime_t, inet::W>(
            SimTime(18, SimTimeUnit::SIMTIME_MS), inet::W(-5) ));
    powerChangesBase.insert(std::pair<simtime_t, inet::W>(
            SimTime(20, SimTimeUnit::SIMTIME_MS), inet::W(-3) ));
    auto testPowerChanges = new std::map<simtime_t, inet::W>(powerChangesBase);
    inet::physicallayer::ScalarNoise test(startTime, endTime, centerFrequency, bandwidth, testPowerChanges);

    ASSERT(test.computeMaxPower(startTime, endTime) == inet::W(8) ); // Fails: result == W(0)
```

The result should clearly be `W(8)`. This happens whenever the "test interval" is shorter than any `powerChanges` gap, as well as some other boundary cases. If there is power change at a boundary change it works ok.

## Severity/Prevelance

* Is is common for such short "test intervals" because of `Radio::isListeningPossible()` : `const IListening *listening = receiver->createListening(this, now, now + 1E-12, position, position);`
* Most of the time `isListeningPossible` is called at the same time as receptionEnd or receptionStart so there is a powerChange `now`. So it still calculates this Ok
* If there is radio transition to `RECEIVER` or `TRANSCEIVER` during an ongoing radio signal, then there is no radio power change at `now` but there will be one from `...BackgroundNoise::computeNoise(...)`. Since the noise is modelled by a signal being added between listeningStart and listeningEnd. So it still calculates this Ok
* If there is no background noise module. Then there will not be a power change signal at the "test interval" bounds, and almost certainly not be one between the bounds. 
* Radio transitions to listening while others are in the middle of transmissions is more common than initially expected because of contention during acknowledgments.

## Possible fix

```
computeMaxPower(simtime_t startTime, simtime_t endTime) const
{
    W noisePower = W(0);
    W maxNoisePower = W(0);
    simtime_t previousStart = 0;
    for (const auto & elem : *powerChanges) {
        if(endTime < elem.first){
            // Power change after end time
            if (std::isnan(maxNoisePower.get()) || previousStart < startTime)
                // Previous power change was before start time, set to current power
                maxNoisePower = noisePower;
            break;
        }

        noisePower += elem.second;
        previousStart = elem.first;
        if ((std::isnan(maxNoisePower.get()) || noisePower > maxNoisePower) && startTime <= elem.first && elem.first < endTime)
            maxNoisePower = noisePower;
    }
```

## Longer term fix
Should it use the same integration method as used in `DimensionalNoiseModel`? 

It would be more accurate for the the background noise to start at simtime 0 and noise end at simtime::Max. Currently the noise is only "present" in some senses whilst there is a "Listening" happening. What is the point of "limiting the noise" to the listening time as in `INoise::getBackgroundNoise()` and `...AnalogModelBase::addNoise(...)`

**Comment by levy** [comment]
Hmm, the very good bug description, thanks!

I think we should avoid the dimensional noise representation and the corresponding multi-dimensional math package if possible. Even if it would perhaps very nicely solve this 1D issue but it has it's own complexity and I think it's difficult to understand for the average user. I didn't check your fix yet, but I guess something like that should be used.

As for the limiting the noise question, I don't quite remember. Maybe, it's just some left-over from a previous era.

**Comment by levy** [comment]
I think this bug has been fixed with merging in #701 so I close this issue.
